django/db/migrations/operations/fields.py

```diff
@@ -63,7 +63,7 @@ def references_field(self, model_name, name, app_label=None):
     def reduce(self, operation, app_label=None):
         return (
             super().reduce(operation, app_label=app_label) or
-            not operation.references_field(self.model_name, self.name, app_label)
+            operation.references_field(self.model_name, self.name, app_label)
```

Comment:
I'm not sure I understand the rationale behind this change, could you elaborate on it? Was it only swapped because it happened to make the tests pass? It's possible this line wasn't even required previously.

The `reduce` signature is `Union[List[Operation], bool]` where `bool` means _should the optimization be allowed to process through_. Why should the optimization made more _lax_ if an operation refers to the field this operation operates on and _stricter_ if it's not the case?

Reply:
I think the _through_ test failure has more to do with how `RemoveField.references_field` deals with through. For example, not saying this is the right solution here, but the following diff happens to make the tests pass as well

```diff
diff --git a/django/db/migrations/operations/fields.py b/django/db/migrations/operations/fields.py
index 41c389f79f..d6cbd6b9c6 100644
--- a/django/db/migrations/operations/fields.py
+++ b/django/db/migrations/operations/fields.py
@@ -36,7 +36,7 @@ class FieldOperation(Operation):
             return field_references_model(self.field, ModelTuple(app_label, name_lower))
         return False

-    def references_field(self, model_name, name, app_label=None):
+    def references_field(self, model_name, name, app_label=None, reference_through=True):
         model_name_lower = model_name.lower()
         # Check if this operation locally references the field.
         if model_name_lower == self.model_name_lower:
@@ -53,11 +53,12 @@ class FieldOperation(Operation):
                         (not hasattr(self.field, 'to_fields') or
                             name in self.field.to_fields or None in self.field.to_fields)):
                     return True
-                through = getattr(remote_field, 'through', None)
-                if (through and ModelTuple.from_model(through) == model_tuple and
--- a/django/db/migrations/operations/fields.py
+++ b/django/db/migrations/operations/fields.py
@@ -36,7 +36,7 @@ class FieldOperation(Operation):
             return field_references_model(self.field, ModelTuple(app_label, name_lower))
         return False

-    def references_field(self, model_name, name, app_label=None):
+    def references_field(self, model_name, name, app_label=None, reference_through=True):
         model_name_lower = model_name.lower()
         # Check if this operation locally references the field.
         if model_name_lower == self.model_name_lower:
@@ -53,11 +53,12 @@ class FieldOperation(Operation):
                         (not hasattr(self.field, 'to_fields') or
                             name in self.field.to_fields or None in self.field.to_fields)):
                     return True
-                through = getattr(remote_field, 'through', None)
-                if (through and ModelTuple.from_model(through) == model_tuple and
-                        (getattr(remote_field, 'through_fields', None) is None or
-                            name in remote_field.through_fields)):
-                    return True
+                if reference_through:
+                    through = getattr(remote_field, 'through', None)
+                    if (through and ModelTuple.from_model(through) == model_tuple and
+                            (getattr(remote_field, 'through_fields', None) is None or
+                                name in remote_field.through_fields)):
+                        return True
         return False

     def reduce(self, operation, app_label=None):
@@ -186,6 +187,11 @@ class RemoveField(FieldOperation):
     def describe(self):
         return "Remove field %s from %s" % (self.name, self.model_name)

+    def references_field(self, model_name, name, app_label=None):
+        return super().references_field(
+            model_name, name, app_label=app_label, reference_through=False
+        )
+
     def reduce(self, operation, app_label=None):
         from .models import DeleteModel
         if isinstance(operation, DeleteModel):
```

Reply:
I don't quite follow. The `references_field()` returns `False` if it's not able to find a field that the operation references to. So if the value of the said `bool` is `False`, doesn't that make the optimization stricter?

Reply:
The previous return value was `not operation.references_field` which meant

1. `not True -> False`, if the operation refers to the field block optimizations through.
2. `not False -> True`, if the operation doesn't refer to the field allow optimizations through.

You proposed change makes it the other way around.

Reply:
Okay, this is my mistake, I got confused with the logic. Removing the `not`, made all the tests pass, so I thought otherwise. I'll be happy to make the proposed changes and fix this.